File: frontend/types/system.ts
import { z } from "zod";
import { GasTypeSchema } from "./common";

export const SystemSettingsSchema = z
  .object({
    id: z.string(),
    is_setup_completed: z.boolean(),
    currency_code: z.string(),
    money_decimals: z.number(),
    created_at: z.string(),
  })
  .passthrough();
export type SystemSettings = z.infer<typeof SystemSettingsSchema>;

export const SystemInitializeInputSchema = z.object({
  sell_price_12: z.number(),
  sell_price_48: z.number(),
  buy_price_12: z.number().optional(),
  buy_price_48: z.number().optional(),
  sell_iron_price_12: z.number().optional(),
  sell_iron_price_48: z.number().optional(),
  buy_iron_price_12: z.number().optional(),
  buy_iron_price_48: z.number().optional(),
  full_12: z.number(),
  empty_12: z.number(),
  full_48: z.number(),
  empty_48: z.number(),
  cash_start: z.number(),
  company_payable_money: z.number().optional(),
  company_full_12kg: z.number().optional(),
  company_empty_12kg: z.number().optional(),
  company_full_48kg: z.number().optional(),
  company_empty_48kg: z.number().optional(),
  currency_code: z.string().optional(),
  money_decimals: z.number().optional(),
  customer_debts: z
    .array(
      z.object({
        customer_id: z.string(),
        money: z.number().optional().default(0),
        cyl_12: z.number().optional().default(0),
        cyl_48: z.number().optional().default(0),
      })
    )
    .optional(),
});
export type SystemInitializeInput = z.infer<typeof SystemInitializeInputSchema>;

export const LedgerHealthIssueSchema = z.object({
  issue_type: z.enum(["mismatch", "orphan"]),
  source_type: z.string(),
  source_id: z.string(),
  message: z.string(),
});
export type LedgerHealthIssue = z.infer<typeof LedgerHealthIssueSchema>;

export const SystemHealthCheckSchema = z.object({
  ok: z.boolean(),
  checked_at: z.string(),
  mismatches: z.number(),
  orphans: z.number(),
  issues: z.array(LedgerHealthIssueSchema).optional().default([]),
});
export type SystemHealthCheck = z.infer<typeof SystemHealthCheckSchema>;

export const SystemSchema = z
  .object({
    id: z.string(),
    customer_id: z.string(),
    name: z.string(),
    gas_type: GasTypeSchema,
    note: z.string().nullish(),
    requires_security_check: z.boolean().optional(),
    security_check_exists: z.boolean().optional(),
    last_security_check_at: z.string().nullish(),
    next_security_check_at: z.string().nullish(),
    is_active: z.boolean().optional(),
    created_at: z.string().nullish(),
  })
  .passthrough();
export type System = z.infer<typeof SystemSchema>;

export const SystemCreateInputSchema = z.object({
  customer_id: z.string(),
  name: z.string(),
  gas_type: GasTypeSchema,
  note: z.string().nullish().optional(),
  requires_security_check: z.boolean().optional(),
  security_check_exists: z.boolean().optional(),
  last_security_check_at: z.string().nullish().optional(),
  is_active: z.boolean().optional(),
});
export type SystemCreateInput = z.infer<typeof SystemCreateInputSchema>;

export const SystemUpdateInputSchema = SystemCreateInputSchema.partial();
export type SystemUpdateInput = z.infer<typeof SystemUpdateInputSchema>;

export const SystemTypeOptionSchema = z
  .object({
    id: z.string(),
    name: z.string(),
    is_active: z.boolean().optional(),
    created_at: z.string().nullish(),
  })
  .passthrough();
export type SystemTypeOption = z.infer<typeof SystemTypeOptionSchema>;

File: backend/app/schemas/system.py
from __future__ import annotations

from datetime import date, datetime
from typing import Literal, Optional

from sqlmodel import Field, SQLModel

from .common import GasType


class SystemCreate(SQLModel):
  customer_id: str
  name: str
  gas_type: GasType
  note: Optional[str] = None
  requires_security_check: bool = False
  security_check_exists: bool = False
  last_security_check_at: Optional[date] = None
  is_active: bool = True


class SystemUpdate(SQLModel):
  customer_id: Optional[str] = None
  name: Optional[str] = None
  gas_type: Optional[GasType] = None
  note: Optional[str] = None
  requires_security_check: Optional[bool] = None
  security_check_exists: Optional[bool] = None
  last_security_check_at: Optional[date] = None
  is_active: Optional[bool] = None


class SystemOut(SQLModel):
  id: str
  customer_id: str
  name: str
  gas_type: GasType
  note: Optional[str] = None
  requires_security_check: bool = False
  security_check_exists: bool = False
  last_security_check_at: Optional[date] = None
  next_security_check_at: Optional[date] = None
  is_active: bool = True
  created_at: datetime


class SystemTypeOptionCreate(SQLModel):
  name: str


class SystemTypeOptionUpdate(SQLModel):
  name: Optional[str] = None
  is_active: Optional[bool] = None


class SystemTypeOptionOut(SQLModel):
  id: str
  name: str
  is_active: bool
  created_at: datetime


class SystemSettingsOut(SQLModel):
  id: str
  is_setup_completed: bool
  currency_code: str
  money_decimals: int
  created_at: datetime


class CustomerOpeningBalance(SQLModel):
  customer_id: str
  money: int = 0
  cyl_12: int = 0
  cyl_48: int = 0


class SystemInitialize(SQLModel):
  sell_price_12: int
  sell_price_48: int
  buy_price_12: int = 0
  buy_price_48: int = 0
  sell_iron_price_12: int = 0
  sell_iron_price_48: int = 0
  buy_iron_price_12: int = 0
  buy_iron_price_48: int = 0
  company_iron_price_12: int = 0
  company_iron_price_48: int = 0
  full_12: int
  empty_12: int
  full_48: int
  empty_48: int
  cash_start: int
  company_payable_money: int = 0
  company_full_12kg: int = 0
  company_empty_12kg: int = 0
  company_full_48kg: int = 0
  company_empty_48kg: int = 0
  currency_code: Optional[str] = None
  money_decimals: Optional[int] = None
  customer_debts: Optional[list[CustomerOpeningBalance]] = None


class LedgerHealthIssue(SQLModel):
  issue_type: Literal["mismatch", "orphan"]
  source_type: str
  source_id: str
  message: str


class SystemHealthCheckOut(SQLModel):
  ok: bool
  checked_at: datetime
  mismatches: int
  orphans: int
  issues: list[LedgerHealthIssue] = Field(default_factory=list)
